### ISSUE TITLE
Job keeps retrying (and failing) every minute when scheduled for every 15 minutes

### ISSUE THREAD

**Issue by Fossil01** [laravel/framework#30264]
- Laravel Version: 6.1.0
- PHP Version: 7.3.9
- Database Driver & Version: MariaDB 10.4.8

### Description:
One of the Jobs defined in Kernel.php keeps retrying every minute even when the first run was successful and the Job was scheduled to run every 15 mins in Kernel.php

```php
$schedule->job(new GeneratePopularDownloadStats())->everyFifteenMinutes()->runInBackground();
```

### Steps To Reproduce:
See this thread: https://old.reddit.com/r/laravel/comments/dhax10/scheduled_job_keeps_failing_and_retrying_but_when/

This all happens on my Homestead VM, all latest version.

**Job:**
```php
<?php

namespace App\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use App\Models\PopularReleases;
use Illuminate\Support\Facades\DB;

class GeneratePopularDownloadStats implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $tries = 3;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        $tvStats = DB::table('download_stats as dl')->select('r.id', 'r.searchname', 'r.guid', 'r.categories_id', 'c.title', 'rc.title as root_category', 'r.grabs', 'dl.count')->join('releases as r', 'dl.release_id', '=', 'r.id')->join('categories as c', 'r.categories_id', '=', 'c.id')->join('root_categories as rc', 'c.root_categories_id', '=', 'rc.id')->whereBetween('r.categories_id', [5000, 5999])->limit(100)->get();

        foreach ($tvStats as $tv) {
            $popular = PopularReleases::updateOrCreate(['release_id' => $tv->id]);
            $popular->release_id = $tv->id;
            $popular->searchname = $tv->searchname;
            $popular->guid = $tv->guid;
            $popular->category = $tv->root_category . ' > ' . $tv->title;
            $popular->type = 2;
            $popular->grabs = $tv->grabs;
            $popular->count = $tv->count;
            $popular->save();
        }
    }
}

```

The error I keep getting when it fails is:
`App\Jobs\GeneratePopularDownloadStats has been attempted too many times or run too long. The job may have previously timed out.`

Which makes sense because somehow it keeps being retried. I flushed Redis (which is my queue driver), truncated failed_jobs table and started queue:work again. This happened:

```
[2019-10-13 20:15:06][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:15:06][1] Processed:  App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:16:36][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:16:36][1] Processed:  App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:18:06][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:18:06][1] Processed:  App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:19:36][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:19:36][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:21:06][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:21:07][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:22:37][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:22:37][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:24:07][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:24:07][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:25:37][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:25:37][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:27:07][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:27:07][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:28:37][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:28:37][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:30:04][2] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:30:04][2] Processed:  App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:30:07][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:30:07][1] Failed:     App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:31:34][2] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:31:34][2] Processed:  App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:31:37][1] Processing: App\Jobs\GeneratePopularDownloadStats
[2019-10-13 20:31:37][1] Failed:     App\Jobs\GeneratePopularDownloadStats
```

Easier to see as an image: 
![screenshot](https://user-images.githubusercontent.com/8832687/66720181-c50b4300-edf9-11e9-8dcb-6286d0f31dcc.png)

So the first run was OK and right on the 15 min mark of the hour. Then after a minute it for some reason decides to run again, and again the next minute. After that it keeps being run every min but fails with the error shown above.

BUT on the next full quarter of the hour (20:30) the job completes successfully again? And then it fails and then it's successful again after which it keeps failing. What the actual fuck?

I am really lost on this one. How does it even run multiple times per minute when the cronjob for the scheduler only runs every minute?

**Comment by Fossil01** [comment]
Ok so with a little help from IRC we figured out that:

A). Don't Google Laravel docs, you will end up on an older version of the docs.
B). The 6.x version of the docs state:

>  The runInBackground method may only be used when scheduling tasks via the command and exec methods.

So I removed that, flushed redis, truncated failed_jobs, started the queue worker and at 22:00 the job succeeded. Then at 22:06 that same job had the status failed...

**Comment by Fossil01** [comment]
Removed withoutOverlapping() too. I have retry_after set to 90s in queue.php

Now:
- Job runs fine, completes without error
- 90s later that same Job suddenly gets the status FAILED.  The runtime for the Job also changes from 0.3s to 90 seconds.

Beats me. I am giving up on this crap and switching to a scheduled artisan command.

**Comment by brunogaspar** [comment]
By the looks of it, it can be a race condition problem, but i can be wrong as i don't know your entire environment/setup.

Trying to help a little bit, have you checked why it's failing, by for example checking the Laravel or even PHP logs for any potential errors?

Also, try adding some logging into the job and see where it might fail or where it might not trigger the logging call.

There's quite a few things you can still do, so don't give up.

**Comment by Fossil01** [comment]
Hi,

Yeah I have checked all the logs, nothing. Just the error that the job was tried too many times. I've added a failed() method to the job and logged the exception but nothing fails. Even when I run the job with just a simple echo or whatever in there it behaves the same.

**Comment by brunogaspar** [comment]
hmm that's odd

My second guess would be that the job was timing out for some reason, but since you say that even with the job doing nothing or with that `echo` it still fails, must be something else.

Have you tried to reproduce the same behaviour on a clean installation of Laravel + Horizon and a simple job but with the same or similar Horizon settings?

**Comment by Fossil01** [comment]
I have referenced another issue. Today I noticed my other Jobs (non scheduled) fail too after initially succeeding.

I have a job which sends contact form emails, the job will succeed and is shown like so in Horizon. I also receive the corresponding e-mail and then after 60-90s that same job fails and keeps being retried.

I've spoken to others now on IRC who have the same issue as me. And then there is the Horizon issue I linked.

Since this also happens when just running queue:work I think it's not a horizon issue but a framework issue.

**Comment by satyago** [comment]
@Fossil01 are you running your queue off of Redis or the database?

We're not running on Laravel 6.1.0 just yet for our software, but I can say that I personally am not seeing this with Laravel 5.8.
However, we did see a lot of quite severe (for us) bugs with MariaDB in the recent weeks, which made us switch back to MySQL 8. So it would be interesting to know which queue driver you've chosen and if switching it makes any difference.

**Comment by Fossil01** [comment]
@satyago Redis. I am now trying to downgrade from 6.1 to 6.0.4 and see what happens.

**Comment by Fossil01** [comment]
Downgraded to 6.0.4, flushed failed_jobs and redis and ran queue:work again. Same issue :-(

**Comment by satyago** [comment]
@Fossil01 I've somehow managed to ignore your Redis mentions, stupid.
Both downgrading and switching the queue driver should be good tests (IMHO) to see if this is setup specific.

And with Horizon setup I've never run `php artisan queue:work`, I've always used `php artisan horizon` as the docs state https://laravel.com/docs/6.x/horizon#running-horizon

This then works off of the horizon config.

**Comment by Fossil01** [comment]
@satyago Yeah, I normally run Horizon using supervisor but for debugging now I wanted to know if the issue also happened when just using the normal queue worker, which it does.

I'll change the queue driver and have a new go at it.

**Comment by Fossil01** [comment]
Okay. So when using the sync driver everything works (as expected). Using the database driver everything seems to be working too.

Now that I think of it, I installed the laravel-lodash package last week to be able to run Redis with IGBinary. I don't think the issue started right after installing this though but I try what happens when I remove it.

**Comment by satyago** [comment]
That sounds like progress.

I also see a few other flaws in the Job code itself, which could end you up in race conditions, or just unnecessary edits.

See, you're limiting the query to 100 results, but you're not preventing this job from running more than once at the same time and you're not keeping track of which entries you're editing.
It could occasionally grab the same entries and then attempt updates or / inserts on those with the `updateOrCreate()` call to the DB, while another process is already processing them.
In other words, it is quite likely the job is running into DB based locks (which with InnoDB is happening on the row level now), but you could also run into "too many open connections" errors and other things if this is done at scale - with luck you're not going to run into this too soon, but for me it'd be something I'd not be willing to risk. It is very difficult to pinpoint errors when the DB starts misbehaving, so I'd optimize code to prevent issues like this first.

**Comment by Fossil01** [comment]
I had the withoutoverlapping() set before, removed it for debugging. But yes you're totally right. 

In the mean time I had switched from a Job to a scheduled artisan command which worked fine. But this morning I noticed other jobs had the same issue.

I have disabled the lodash serviceproviders just now, ran `artisan horizon` and it looks like everything works now. lol, never thought this would be the culprit.

The one weird thing though is that `artisan queue:work` doesnt run my `contactemail` queue unless I specify it:  `artisan queue:work --queue=contactemail`

**Comment by satyago** [comment]
I think the intent here is that you could run multiple workers for specific queues.
At least it's always been the case that it only works off of the `default` queue if you don't specify another.

**Comment by Fossil01** [comment]
Ah, yeah. I remember it running all queues but my brain is so chaotic because of all the debugging that I'm most probably wrong.

Thanks for the help btw! Hopefully this issue helps others who try to run Redis/Laravel with IGBinary (with or without laravel-lodash).

**Comment by driesvints** [comment]
@Fossil01 seems like you managed to solve your issue 👍 

Always feel free to send a pr to the docs if you feel like something can be explained more thoroughly.

**Comment by Fossil01** [comment]
Cheers, sorry forgot to close the issue.

**Comment by bunea** [comment]
We're seeing this issue as well even with predis. After a couple of days of debugging and trying out different things we realized that `retry_after` was retrying a running job. For the time being we increased that value to be over the job's timeout.

Maybe it's worth reopening this.

**Comment by raymondtri** [comment]
> We're seeing this issue as well even with predis. After a couple of days of debugging and trying out different things we realized that `retry_after` was retrying a running job. For the time being we increased that value to be over the job's timeout.
> 
> Maybe it's worth reopening this.

I'm still noticing this issue as well.

**Comment by akalongman** [comment]
Queue just not working when Redis uses igbinary, so I use another non-igbinary connection for queue itself
```
        // Queue Redis connection
        'queue'   => [
            'host'     => env('QUEUE_REDIS_HOST', '127.0.0.1'),
            'password' => env('QUEUE_REDIS_PASSWORD', null),
            'port'     => env('QUEUE_REDIS_PORT', 6379),
            'database' => 0,
        ],

        // Horizon Redis connection
        'horizon' => [
            'host'         => env('QUEUE_REDIS_HOST', '127.0.0.1'),
            'password'     => env('QUEUE_REDIS_PASSWORD', null),
            'port'         => env('QUEUE_REDIS_PORT', 6379),
            'read_timeout' => 3,
            'serializer'   => 'igbinary',
        ],
```
